Sources/_Datadog_Private/ObjcAppLaunchHandler.m

```diff
@@ -4,60 +4,94 @@
+// https://developer.apple.com/documentation/uikit/app_and_environment/responding_to_the_launch_of_your_app/about_the_app_launch_sequence
 
-@interface AppLaunchTimer : NSObject
-@property NSDate *frameworkLoadTime;
-// Knowing how this value is collected and that it is written only once, the use of
-// `atomic` property is enough for thread safety. It guarantees that neither partial
-// nor garbage value will be returned.
-@property (atomic) NSTimeInterval timeToApplicationBecomeActive;
+// Note that the following properties will be set early enough in the App lifecycle to avoid any race conditions without any specific synchronization means.
+static NSTimeInterval TimeToApplicationFinishedLaunching = 0.0;
+static NSTimeInterval TimeToLoadApplicationRootUI = 0.0;
+
+@interface AppLaunchHandler : NSObject
 @end
 
-@implementation AppLaunchTimer
+@implementation AppLaunchHandler {
+    NSTimeInterval _processStartTime;
+    NSTimeInterval _applicationMainEndTime;
+    NSTimeInterval _frameworkLoadTime;
+}
+
++ (void)load {
+    // This is called at the `_Datadog_Private` load time, keep the work minimal
+    NSTimeInterval now = CFAbsoluteTimeGetCurrent();
+    // Schedule a call for the first turn of run loop before the main nib/storyboard is loaded:
+    [NSRunLoop.mainRunLoop performInModes:@[NSRunLoopCommonModes] block:^{
+        AppLaunchHandler *handler = [[self alloc] initWithFrameworkLoadTime:now];
+        [handler markEndOfUIApplicationMain];
+        [handler scheduleUIApplicationDidFinishLaunchingMarking];
+    }];
+}
 
-- (instancetype)init
+- (instancetype)initWithFrameworkLoadTime:(NSTimeInterval)frameworkLoadTime
 {
-    self = [super init];
-    if (self) {
-        NSDate *time = [NSDate new];
-        self.frameworkLoadTime = time;
-        self.timeToApplicationBecomeActive = -1;
-        [[NSNotificationCenter defaultCenter] addObserver:self
-                                                 selector:@selector(handleDidBecomeActiveNotification)
-                                                     name:UIApplicationDidBecomeActiveNotification
-                                                   object:nil
-         ];
+    if (self = [super init]) {
+        _frameworkLoadTime = frameworkLoadTime;
     }
     return self;
 }
 
-- (void)handleDidBecomeActiveNotification {
-    if (self.timeToApplicationBecomeActive > -1) { // sanity check & extra safety
-        return;
-    }
+- (void)markEndOfUIApplicationMain {
+    _applicationMainEndTime = CFAbsoluteTimeGetCurrent();
+}
 
-    NSDate *time = [NSDate new];
-    NSTimeInterval duration = [time timeIntervalSinceDate:self.frameworkLoadTime];
-    self.timeToApplicationBecomeActive = duration;
-    [[NSNotificationCenter defaultCenter] removeObserver:self];
+- (void)scheduleUIApplicationDidFinishLaunchingMarking {
+    id __block token = [NSNotificationCenter.defaultCenter
+                        addObserverForName:UIApplicationDidFinishLaunchingNotification
```

Comment:
We've been using `UIApplicationDidBecomeActiveNotification` before, but now we use `UIApplicationDidFinishLaunchingNotification` - if they do not mark the same moment in time, this changes the definition of RUM "launch time" for iOS.

How is `UIApplicationDidFinishLaunchingNotification` different than the first `UIApplicationDidBecomeActiveNotification`?

Reply:
After the `UIApplication.delegate` is done with the callbacks and `UIWindow` is made key and visible, the sequence of notifications will be `UIApplicationDidFinishLaunchingNotification` then `UIApplicationDidBecomeActiveNotification`. So I believe they happen close to each other (depending on potential observers and the work they do).

That being said `UIApplicationDidBecomeActiveNotification` is a repeating notification that is meant for the whole app lifecycle (although by removing the observer this is only fires initially) and `UIApplicationDidFinishLaunchingNotification` fits with the semantic of App Launch Time better.

I don't think it really changes the definition of RUM "launch time" but I'm ok using `UIApplicationDidBecomeActiveNotification` if that is safer as it's still related to the app first appearance anyway.

Reply:
Got it 👍. I also did some digging around and realised that "finished launching" relates to cold start, whereas "did become active" is for returning from background. Both should work fine and mean the same in our implementation 👌.